### ISSUE TITLE
Real Time Mode

### ISSUE THREAD

**Issue by marcioAlmada** [marcioAlmada/yay#11]
Should allow to activate pre processing of every included file.
- [x] A stream wrapper should be available for single file inclusion:
  
  ``` php
  include 'yay://some/file.php';
  ```
- [ ] Composer integration. Allowing files to be preprocessed automatically during inclusion.
- [ ] Try to take advantage of opcache
  - [ ] It looks like the [stream-is-cacheable](https://wiki.php.net/rfc/streams-is-cacheable) RFC is needed to achieve the task above, so... Help to pass the [stream-is-cacheable](https://wiki.php.net/rfc/streams-is-cacheable) RFC

**Comment by sekjun9878** [comment]
Hello, I've made an attempt at writing a composer plugin for Real-Time mode integration with Yay, and it seems that all that needs to be done is a plugin that patches ClassLoader.php's includeFile function to simply prepend the yay:// wrapper before the filename?

**Comment by marcioAlmada** [comment]
Yes, that's one of the possibilities. Patching ClassLoader.php is little too ad hoc to my taste and probably won't last long.

The other idea is to hijack the composer autoloader instance and force everything to be included through `yay://`, but I havent checked if it's possible yet (no time):

```php
$autoloader = include 'vendor/autoload.php';
// do some sourcery with $autoloader instance
```

And there is also the more formal approach to ask composer project for a configuration or an API change on autoloader that allows registering a wrapper for inclusion.

**Comment by marcioAlmada** [comment]
Also, for the real time mode to work decently we would need the [stream-is-cacheable](https://wiki.php.net/rfc/streams-is-cacheable) RFC to pass. So, any support to this RFC is welcome too :)

**Comment by sekjun9878** [comment]
Hello

I've looked at the composer autoloader and composer plugins only work the cli level, not at the autoloader level so the only way to overload the autoloader is to subscribe to the post-autoloader-dump hook and modify the autoloader after dumping is finished. So I think patching the ClassLoader is the only way since ClassLoader is where the include statement is at and ClassLoader is copy-pasted, not generated so no modification of Composer instance would change anything in ClassLoader unless we edit the file directly.

**Comment by sekjun9878** [comment]
It's also possible to simply have a cache directory, where the code is dumped postprocessed, and the Composer autoloader can be re-wired to include from that directory as a prefix but then it would not be "real-time".

**Comment by TazeTSchnitzel** [comment]
You could make the files themselves have an explicit call to yay, avoiding any magic:

    \yay\compile(__FILE__); die();

**Comment by marcioAlmada** [comment]
@TazeTSchnitzel you genious  :smirk:

I still want composer integration but that's indeed a really cool idea.

**Comment by TazeTSchnitzel** [comment]
:D

**Comment by TazeTSchnitzel** [comment]
I just realised that this wouldn't work if what followed wasn't valid PHP syntax.

But you could use `__halt_compiler` to mitigate that.

**Comment by marcioAlmada** [comment]
Yes of course, otherwise it would fail upon parsing. I'm building a sample bootstrap project to test all possibilities.

**Comment by sekjun9878** [comment]
@TazeTSchnitzel You sir, are a genius! Thanks for that solution.

**Comment by marcioAlmada** [comment]
it... works https://github.com/marcioAlmada/yay-enums/blob/master/enum.tests.php :manic_laugh:

**Comment by marcioAlmada** [comment]
We are going to use a small php extension for engine integration instead. See #32.

**Comment by CMCDragonkai** [comment]
Can the macros be executed on a function execution, I was looking for something like lisp-like macros.

**Comment by sunel** [comment]
Why cant we do this in the entry file?

```php
$autoload = require __DIR__.'/vendor/autoload.php';

spl_autoload_unregister(array($autoload, 'loadClass'));

function includeFile($file)
{
	include $file;
	echo "CALLED";
}

spl_autoload_register(function($class) use ($autoload) {
	if ($file = $autoload->findFile($class)) {
			includeFile($file);
			return true;
	}
}, true, true);
```

**Comment by chris-kruining** [comment]
@sunel I see a decision has been made, see #32 ;)

**Comment by marcioAlmada** [comment]
@sunel https://github.com/preprocess - uses yay as its base -  and If I'm not mistaken it was made by modifying composer autoload.

**Comment by sunel** [comment]
@marcioAlmada ya it uses YAY "preprocess" uses a different file extension "pre".

**Comment by assertchris** [comment]
@sunel this is so that the preprocessing is opt-in per file. Could look at configuration to find some other way to opt-in and some other scheme for the file creation, but I think it will always be opt-in per file.

**Comment by marcioAlmada** [comment]
```
<?php declare(use_preprocess=1);
```

This is indistinguishable from trolling, I'll let you decide :sweat_smile:

**Comment by xtlsoft** [comment]
There is an example on how to hook composer's autoloader: https://github.com/ircmaxell/PhpGenerics/blob/master/lib/bootstrap.php

**Comment by grikdotnet** [comment]
hey, the blocker is code in OpCache - a hardcoded list of streams that are allowed to be cached when used in include(), exactly "file" and "phar"

I hacked this with a custom stream wrapper and autoloader. It uses Composer to locate files. Here: is the code: https://github.com/grikdotnet/generics/tree/master/generics/Internal/service
The "file" stream can be temporarily redefined, and it's fast.

It's a dirty hack, though, but seems working, Maybe it will help.

**Comment by grikdotnet** [comment]
> [@sunel](https://github.com/sunel) I wouldn't personally under any circumstance bypass/overload the composer autloading. I do believe composer has some 'event'-based system for autloading plugins, I think it is better to utilize that feature, imho

the Composer autoloader is just the include() call, nothing special
Multiple autoloaders for composer is a legal, documented feature, and used in multiple packages
